### ISSUE TITLE
Components only render after a hot reload

### ISSUE THREAD

**Issue by osharaki** [flame-engine/flame#541]
## Description

I'm following along with the joystick example [here](https://github.com/flame-engine/flame/tree/master/doc/examples/joystick). When I attempt a hot restart of my app, this is what I see:

![image](https://user-images.githubusercontent.com/38385026/100088516-8b0bcd80-2e50-11eb-8555-b0b2af558eba.png)

Only when I trigger a hot reload do I actually see the components being rendered properly:

![image](https://user-images.githubusercontent.com/38385026/100088677-cd350f00-2e50-11eb-92d2-4a4d02cd42fa.png)

## Development environment

Flame version: ^1.0.0-rc2

Flutter doctor:

```
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel stable, 1.22.3, on Microsoft Windows
    [Version 10.0.19041.630], locale en-US)
 
[√] Android toolchain - develop for Android devices (Android SDK
    version 30.0.2)
[!] Android Studio (version 4.1.0)
    X Flutter plugin not installed; this adds Flutter specific  
      functionality.
    X Dart plugin not installed; this adds Dart specific        
      functionality.
[!] IntelliJ IDEA Community Edition (version 2019.3)
    X Flutter plugin not installed; this adds Flutter specific  
      functionality.
    X Dart plugin not installed; this adds Dart specific
      functionality.
[√] VS Code, 64-bit edition (version 1.51.1)
[√] Connected device (1 available)    

! Doctor found issues in 2 categories.
```

## Runtime

This issue is related to running on which platform? (Select all that apply)

 - [x] Android :android:
 - [ ] iOS :apple:
 - [ ] Web :browser:

## Minimal reproducible code (Required for bugs)

A minimal reproducible example can be found [here](https://github.com/osharaki/game_off_2020/tree/component-render-error-demo/lib).

**Comment by spydon** [comment]
How are you running it?

**Comment by osharaki** [comment]
In VS Code I hit Ctrl+F5 (Run without debugging).

**Comment by spydon** [comment]
Do you get the same result if you just run `flutter run` from the terminal?

**Comment by osharaki** [comment]
Yes. Then if I enter r (hot reload), it works as expected, but if I enter R (hot restart), the issue arises.

**Comment by osharaki** [comment]
It seems when I enter s (take a screenshot), this has the same effect as doing a hot reload.

**Comment by osharaki** [comment]
Changing the square from type `Component` to *flame_forge2d*'s `BodyComponent` causes it to render from the start as expected. The problem persists for the joystick, though.

**Comment by erickzanardo** [comment]
@osharaki Can you confirm is this is still happening on the more recent version of Flame? (v1.0.0-rc4)

Also, I noticed that you are adding your components on the constructor of your game, you could also try to add them on the `onLoad` method, like

```dart
  @override
  Future<void> onLoad() {
    joystick.addObserver(player);
    add(player);
    add(joystick);
  }
 ```

**Comment by osharaki** [comment]
When I upgrade to Flame `1.0.0-rc4`, `runApp(MainGame().widget)` now causes an error:
> The getter 'widget' isn't defined for the type 'MainGame'
```yml
dependencies:
  ...
  flame: ^1.0.0-rc4
```

**Comment by spydon** [comment]
@osharaki you have to wrap it in a `GameWidget`.

```dart
runApp(
  GameWidget(
    game: MainGame()
  )
);
```

**Comment by osharaki** [comment]
> @osharaki Can you confirm is this is still happening on the more recent version of Flame? (v1.0.0-rc4)
> 
> Also, I noticed that you are adding your components on the constructor of your game, you could also try to add them on the `onLoad` method, like
> 
> ```dart
>   @override
>   Future<void> onLoad() {
>     joystick.addObserver(player);
>     add(player);
>     add(joystick);
>   }
> ```

After upgrading and applying @spydon's comment, the behavior still hasn't changed. However, when I add the components in `onLoad()` instead, as @erickzanardo suggested, it works fine. I did need to add `return super.onLoad();`, though. So:
```dart
@override
  Future<void> onLoad() {
    joystick.addObserver(player);
    add(player);
    add(joystick);
    return super.onLoad();
  }
```
 In fact, `onLoad()` also fixes the issue on `1.0.0-rc2`. So it seems this was simply a matter of incorrectly adding the components all along?

**Comment by erickzanardo** [comment]
You shouldn't need to return 'super.onLoad()' that probably was required because you didn't added the 'async' keyword to your method declaration.

Since the use of the onLoad method solved the issue, I am gonna close this, feel free to add more questions here if needed.
